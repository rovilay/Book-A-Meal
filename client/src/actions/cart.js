export const addMealToCart = ({
  id,
  title,
  price: unitPrice,
  portion
}) => ({
  type: 'ADD_MEAL_TO_CART',
  meal: {
    id,
    title,
    unitPrice,
    portion
  }
});

export const deleteMealInCart = ({ id }) => ({
  type: 'DELETE_MEAL_IN_CART',
  mealId: id
});

export const emptyCart = () => ({
  type: 'EMPTY_CART',
});

