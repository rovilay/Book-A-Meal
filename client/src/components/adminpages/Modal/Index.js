import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Modal from 'react-modal';

import MenuDetailsTable from '../dashboard/MenuTable/MenuDetails';
import EditMenuTable from '../dashboard/MenuTable/EditMenu';
import OrderDetailsTable from '../orderPage/orderDetails';

Modal.setAppElement('#root');
class ModalComp extends Component {
  componentDidMount() {
    this.props.emptyEditMenu();
  }
  render() {
    const { hideModal, modal } = this.props;
    const {
      isOpen,
      isInfo,
      isEdit,
      isOrderInfo,
      content,
      contentLabel,
    } = modal;

    return (
      <Modal
        isOpen={isOpen}
        contentLabel= {contentLabel}
        overlayClassName="overlay"
        className="modal"
      >
        <div className="closeBtn">
          <button
            onClick={hideModal}
            className="btn-col btn-1"
          >
            close
          </button>
        </div>
        {
          (isInfo && content)
          &&
          (
            <MenuDetailsTable
              title={contentLabel}
              content={content}
              {...this.props}
            />
          )
        }

        {
          (isEdit)
          &&
          (
            <EditMenuTable
              title={contentLabel}
              content={content}
              isEdit={isEdit}
              {...this.props}
            />
          )
        }

        {
          (isOrderInfo)
          &&
          (
            <OrderDetailsTable
              title={contentLabel}
              content={content}
              isEdit={isEdit}
              {...this.props}
            />
          )
        }
      </Modal>
    );
  }
}

ModalComp.defaultProps = {
  emptyEditMenu: undefined
};

ModalComp.propTypes = {
  modal: PropTypes.object.isRequired,
  setModal: PropTypes.func.isRequired,
  hideModal: PropTypes.func.isRequired,
  emptyEditMenu: PropTypes.func,
};

export default ModalComp;
