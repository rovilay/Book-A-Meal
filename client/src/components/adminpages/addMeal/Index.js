import React, { Component } from 'react';
import PropTypes from 'prop-types';

class AddMeal extends Component {
  render() {
    return (
      <div>
        addmeal page
      </div>
    );
  }
}

export default AddMeal;
