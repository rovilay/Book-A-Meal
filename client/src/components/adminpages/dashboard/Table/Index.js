import React, { Component } from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';

import tableHead from '../../../../helpers/tableHead';
import TableHead from '../../../common/Table/TableHead';
import MenuTableRow from './MenuTableRow';

class MenuTable extends Component {
  render() {
    const { menus } = this.props;
    return (
      <div className="menu_table">
        <table>
          <TableHead tableHead={tableHead.menuTableHead} />
          <tbody>
            {
              menus.map((menu, i) => {
                const {
                  id: menuId,
                  postOn,
                  User,
                } = menu;
                const item = {
                  sn: ++i,
                  menuId,
                  postOn: moment(postOn).format('LL'),
                  createdBy: `${User.firstName} ${User.lastName}`
                };
                return (<MenuTableRow
                  key={menuId}
                  item={item}
                />);
              })
            }
          </tbody>
        </table>
      </div>
    );
  }
}

MenuTable.propTypes = {
  menus: PropTypes.array.isRequired
};

export default MenuTable;
