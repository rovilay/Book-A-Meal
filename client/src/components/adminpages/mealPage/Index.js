/*   eslint class-methods-use-this:0, */
/* eslint no-alert: 0 */
/* eslint no-restricted-globals: 0 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { ToastContainer, toast } from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';

import navData from '../../../helpers/navData';
import MealForm from './MealForm';
import MealTable from './MealTable/MealTable';
import adminActions from '../../../actions/admin';
import filterAction from '../../../actions/filter';
import imageUploader from '../../../helpers/imageUploader';
import FilterComp from '../../common/Filter';

class MealPage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isEdit: false,
      mealOnEditId: '',
      imageToUpload: '',
      uploadedImageLink: '',
      disableBtn: false
    };

    this.editMeal = this.editMeal.bind(this);
    this.closeEdit = this.closeEdit.bind(this);
    this.notify = this.notify.bind(this);
    this.fillForm = this.fillForm.bind(this);
    this.clearForm = this.clearForm.bind(this);
    this.getFormVal = this.getFormVal.bind(this);
    this.onAddMeal = this.onAddMeal.bind(this);
    this.onUpdateMeal = this.onUpdateMeal.bind(this);
    this.showUploadBar = this.showUploadBar.bind(this);
    this.uploadImage = this.uploadImage.bind(this);
    this.checkFileSize = this.checkFileSize.bind(this);
  }

  componentDidMount() {
    this.props.setNav(navData.adminNav);
    this.props.getMeals();
    this.props.filterAction('caterer_meals', { filter: 'all' });
    this.closeEdit();
  }

  /**
   * Updates Meal
   *
   * Calls the updateMeal action
   * And close edit state
   */
  onUpdateMeal() {
    const { updateMeal } = this.props;
    const data = this.getFormVal();
    const mealId = this.state.mealOnEditId;
    const confirmed = confirm('Confirm Update!');

    if (confirmed) {
      updateMeal({ mealId, data });
      setTimeout(() => {
        this.notify(this.props.serverRes.message);
        location.reload();
      }, 200);
    }
    this.closeEdit();
  }

  /**
   * Posts Meal
   *
   * Calls the postMeal action
   * Clears form on success
   */
  onAddMeal() {
    const { postMeal } = this.props;
    const data = this.getFormVal();
    const confirmed = confirm('Confirm Add Meal!');
    if (confirmed) {
      postMeal(data);
      setTimeout(() => {
        if (this.props.serverRes.success) {
          this.notify(this.props.serverRes.message);
          this.clearForm();
          location.reload();
        } else {
          this.notify(this.props.serverRes.message);
        }
      }, 200);
    }
  }

  /* eslint prefer-destructuring:0 */
  /**
   * Gets form values and sets image value with link from cloudinary,
   * If no image sets default image
   *
   * @returns {Object} data - Object consist of meal info
   */
  getFormVal() {
    const title = document.getElementById('meal-name').value;
    const price = document.getElementById('price').value;
    const description = document.getElementById('dsc').value;
    let image = this.state.uploadedImageLink;

    if (!image && this.state.isEdit) {
      image = this.props.mealOnEdit.image;
    } else if (!image && !this.state.isEdit) {
      image = 'https://res.cloudinary.com/dcqnswemi/image/upload/v1529300780/default_meal_img.jpg';
    }

    const data = {
      title,
      price,
      description,
      image
    };
    return data;
  }

  /**
   * Shows upload progress bar if image is present
   */
  showUploadBar() {
    if (this.checkFileSize()) {
      const imageFile = document.getElementById('image').files[0];
      this.setState({ imageToUpload: imageFile.name });
    } else {
      this.setState({ imageToUpload: '' });
    }
  }


  /**
   * Uploads image to cloudinary
   */
  uploadImage() {
    this.showUploadBar();
    setTimeout(() => {
      if (this.state.imageToUpload) {
        const url = imageUploader('image');
        if (url) {
          url
            .then((res) => {
              if (typeof (res) === 'string') {
                return this.setState({
                  uploadedImageLink: res,
                  disableBtn: false
                }); // enable btn after upload
              }

              throw new Error(res.message);
            })
            .catch(err => err);
        }
      }
    }, 100);
  }

  /**
   * Sets meal for edit
   *
   * @param {string} mealId Id of meal to edit,
   * Calls setMealForEdit action
   *
   */
  editMeal(mealId) {
    this.setState({ isEdit: true });
    this.props.setMealForEdit(mealId);

    setTimeout(() => {
      this.fillForm();
    }, 200);

    // scroll to top
    document.body.scrollTop = '100px';
    document.documentElement.scrollTop = '100px';
  }

  /**
   * Closes edit state,
   * Calls clearForm function,
   * Dispatch removeMealFromEdit action
   */
  closeEdit() {
    this.setState({
      isEdit: false,
      mealOnEditId: '',
      imageToUpload: '',
      disableBtn: false,
      uploadedImageLink: ''
    });
    this.clearForm();
    this.props.removeMealFromEdit();
  }

  /**
   * Notifies message using toast module
   *
   * @param {*} msg message to notify
   */
  notify(msg) {
    toast(msg, {
      position: toast.POSITION.TOP_CENTER,
      className: 'toast',
      progressClassName: 'toast-progress'
    });
  }

  /**
   * Fills form on edit state with info of meal on Edit,
   * Sets mealOnEditId state
   */
  fillForm() {
    const mealName = document.getElementById('meal-name');
    const price = document.getElementById('price');
    const dsc = document.getElementById('dsc');

    mealName.value = this.props.mealOnEdit.title;
    price.value = this.props.mealOnEdit.price;
    dsc.value = this.props.mealOnEdit.description;
    this.setState({ mealOnEditId: this.props.mealOnEdit.id });
  }

  /**
   * Clears form
   */
  clearForm() {
    const mealName = document.getElementById('meal-name');
    const price = document.getElementById('price');
    const dsc = document.getElementById('dsc');
    const image = document.getElementById('image');

    mealName.value = '';
    price.value = '';
    dsc.value = '';
    image.value = '';
  }

  /**
   * Checks if file size exceeds 1.5mb
   * And disables button if file is proper size
   */
  checkFileSize() {
    const file = document.getElementById('image');
    if (file.files[0]) {
      if (file.files[0].size > 1500000) {
        this.notify('File must not exceed 1.5mb');
        file.value = '';
        return false;
      }
      this.setState({ disableBtn: true });
      return true;
    }

    return false; // if no file
  }

  render() {
    const {
      isEdit,
      mealOnEditId,
      imageToUpload,
      disableBtn
    } = this.state;

    return (
      <div>
        <section className="form-section">
          <div className="meal-container">
            <MealForm
              {...this.props}
              isEdit={isEdit}
              closeEdit={this.closeEdit}
              addMeal={this.onAddMeal}
              updateMeal={this.onUpdateMeal}
              mealOnEditId={mealOnEditId}
              imageToUpload={imageToUpload}
              notify={this.notify}
              showUploadBar={this.showUploadBar}
              uploadImage={this.uploadImage}
              uploadedImageLink={this.state.uploadedImageLink}
              disableBtn={disableBtn}
            />
          </div>
        </section>
        <div className="table-container">
          <div className="table-title">Meal Options</div>
          <FilterComp
            {...this.props}
            tableContent="caterer_meals"
          />
          {
            (this.props.filteredMeals.length === 0)
              ?
                <p className="empty not-found">No meal found!</p>
              :
                <MealTable
                  {...this.props}
                  isEdit={isEdit}
                  editMeal={this.editMeal}
                  notify={this.notify}
                />
          }
        </div>
        <ToastContainer
          {...this.props}
        />
      </div>
    );
  }
}

MealPage.propTypes = {
  setNav: PropTypes.func.isRequired,
  getMeals: PropTypes.func.isRequired,
  setMealForEdit: PropTypes.func.isRequired,
  removeMealFromEdit: PropTypes.func.isRequired,
  mealOnEdit: PropTypes.object.isRequired,
  postMeal: PropTypes.func.isRequired,
  updateMeal: PropTypes.func.isRequired,
  deleteMeal: PropTypes.func.isRequired,
  serverRes: PropTypes.object.isRequired,
  filterAction: PropTypes.func.isRequired,
  filteredMeals: PropTypes.array.isRequired
};

const mapStateToProps = state => ({
  meals: state.admin.meals,
  filteredMeals: state.admin.filteredMeals,
  mealOnEdit: state.admin.mealOnEdit,
  serverRes: state.admin.serverRes,
});

const mapDispatchToProps = dispatch => bindActionCreators(
  {
    ...adminActions,
    filterAction
  },
  dispatch
);

export default connect(mapStateToProps, mapDispatchToProps)(MealPage);
