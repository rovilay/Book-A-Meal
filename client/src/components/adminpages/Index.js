import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import navData from '../../helpers/navData';
import { setNav } from '../../actions/navLinks';


class MenuPage extends Component {
  componentDidMount() {
    this.props.dispatch(setNav(navData.adminNavDefault));
  }
  render() {
    return (
      <div>
        admin Page!
      </div>
    );
  }
}

MenuPage.propTypes = {
  dispatch: PropTypes.func.isRequired,
};

export default connect()(MenuPage);
