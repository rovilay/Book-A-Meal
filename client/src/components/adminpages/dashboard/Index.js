/*
  eslint class-methods-use-this:0,
  eslint no-alert: 0,
*/
/* eslint no-restricted-globals: 0 */
import arraySort from 'array-sort';
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import {
  Accordion,
} from 'react-accessible-accordion';

import navData from '../../../helpers/navData';
import filterify from '../../../helpers/filterify';
import { setNav } from '../../../actions/navLinksActions';
import {
  getTodayMenu, getAllMenus, postMenu,
  updateMenu, addMealToNewMenu, removeMealFromNewMenu,
  emptyNewMenu, emptyEditMenu, addMealInEditMenu,
  deleteMealInEditMenu, setMenuForEdit, getMenuMeals,
  deleteMealInMenu, deleteMenuMeal
} from '../../../actions/menuActions';
import { getMeals } from '../../../actions/mealActions';
import {
  setModal, deleteMealInEditModal,
  addMealInEditMenuModal
} from '../../../actions/modalActions';
import MenuTable from './MenuTable/MenuTable';
import SetMenuCard from './MenuCard/SetMenu';
import Filter from '../../common/Filter';
import setFilter from '../../../actions/filterActions';
import ModalComp from '../Modal/Index';

class AdminDashboard extends Component {
  constructor(props) {
    super(props);

    this.setNewMenuMeal = this.setNewMenuMeal.bind(this);
    this.submitNewMenu = this.submitNewMenu.bind(this);
    this.showMenuDetails = this.showMenuDetails.bind(this);
    this.hideModal = this.hideModal.bind(this);
    this.editMenu = this.editMenu.bind(this);
    this.deleteRow = this.deleteRow.bind(this);
    this.onSubmitUpdate = this.onSubmitUpdate.bind(this);
    this.unCheckAll = this.unCheckAll.bind(this);
    this.getMenus = this.getMenus.bind(this);
    this.handlePaginationClick = this.handlePaginationClick.bind(this);
  }

  componentDidMount() {
    this.props.setNav(navData.adminNavDefault);
    this.props.getMeals({});
    this.props.getAllMenus({});
    this.props.setFilter({ filter: 'all' });
    this.hideModal();
  }

  onSubmitUpdate(menuDate, meals) {
    this.props.updateMenu({ menuDate, meals });
  }

  setNewMenuMeal(mealId) {
    let checkbox;
    if (this.props.modal.isEdit) {
      checkbox = document.getElementById(`${mealId}-edit`);
    } else {
      checkbox = document.getElementById(mealId);
    }
    if (checkbox.checked === true) {
      this.props.addMealToNewMenu(mealId);
    } else if (checkbox.checked === false) {
      this.props.removeMealFromNewMenu(mealId);
    }
  }

  /**
   * gets menus
   * @param {number} limit pagination limit
   * @param {number} offset pagination offset
   * @param {string} menuUrl menu meals url to get meals for a menu
   */
  getMenus({ limit, offset, menuUrl }) {
    this.props.getAllMenus({
      limit,
      offset,
      menuUrl
    });
  }

  hideModal() {
    this.props.setModal({
      isOpen: false,
      isEdit: false,
      isInfo: false,
      close: true,
      contentLabel: '',
      content: {}
    });
  }

  showMenuDetails(menuDetail) {
    this.props.setModal({
      isOpen: true,
      isInfo: true,
      isEdit: false,
      close: false,
      contentLabel: 'Menu details',
      content: { menuDetails: [...menuDetail] }
    });
  }

  editMenu({ menuId, postOn, meals }) {
    getMeals({});
    this.props.setModal({
      isOpen: true,
      isInfo: false,
      isEdit: true,
      close: false,
      contentLabel: 'Add Meals',
      content: { menuId, postOn, meals }
    });
    this.props.setMenuForEdit(meals);
    this.props.emptyNewMenu();
  }

  deleteRow(id) {
    this.props.deleteMealInEditModal(id);
    this.props.deleteMealInEditMenu(id);
  }

  unCheckAll(mealIdArr) {
    mealIdArr.forEach((mealId) => {
      const checkbox = document.getElementById(mealId);
      checkbox.checked = false;
    });
  }

  submitNewMenu() {
    const meals = [...this.props.newMenuMeals];
    const date = document.getElementById('postOn').value;
    const postOn = date.split('/').reverse().join('-');
    this.props.postMenu({ postOn, meals });
  }


  /**
   * handles pagination changes
   *
   * @param {object} data data object from pagination component
   */
  handlePaginationClick(data) {
    // const nextPage = data.selected + 1;
    const { limit } = this.props.pagination;
    const offset = (data.selected) * limit;
    this.getMenus({ limit, offset });
  }

  render() {
    const { user } = this.props;
    const { firstName, lastName } = user;
    return (
      <div>
        <div className="welcome">
          <p className="merienda">
            welcome, {firstName} {lastName}
          </p>
        </div>
        <section className="setmenu">
          <Accordion>
            <div className="setmenu-container">
              <SetMenuCard
                setNewMenuMeal={this.setNewMenuMeal}
                submitNewMenu={this.submitNewMenu}
                {...this.props}
              />
            </div>
          </Accordion>
        </section>
        {/* <MenuAccordion /> */}
        <section className="adminpage">
          <div className="menu-title">Menus List</div>
          <div className="container">
            <Filter
              {...this.props}
              tableContent="Menus_List"
            />
            {
              (this.props.menus.length === 0)
                ?
                  <p className="empty not-found">No menu found!</p>
                :
                  <Accordion>
                    <MenuTable
                      showMenuDetails={this.showMenuDetails}
                      editMenu={this.editMenu}
                      {...this.props}
                    />
                  </Accordion>
            }
          </div>
        </section>
        <ModalComp
          hideModal={this.hideModal}
          deleteRow={this.deleteRow}
          submitUpdate={this.onSubmitUpdate}
          setNewMenuMeal={this.setNewMenuMeal}
          {...this.props}
        />
      </div>
    );
  }
}

AdminDashboard.propTypes = {
  setNav: PropTypes.func.isRequired,
  getMeals: PropTypes.func.isRequired,
  getAllMenus: PropTypes.func.isRequired,
  meals: PropTypes.array.isRequired,
  addMealToNewMenu: PropTypes.func.isRequired,
  removeMealFromNewMenu: PropTypes.func.isRequired,
  postMenu: PropTypes.func.isRequired,
  newMenuMeals: PropTypes.array.isRequired,
  setModal: PropTypes.func.isRequired,
  deleteMealInEditModal: PropTypes.func.isRequired,
  updateMenu: PropTypes.func.isRequired,
  setMenuForEdit: PropTypes.func.isRequired,
  emptyNewMenu: PropTypes.func.isRequired,
  modal: PropTypes.object.isRequired,
  user: PropTypes.object.isRequired,
  menus: PropTypes.array.isRequired,
  setFilter: PropTypes.func.isRequired,
  deleteMealInEditMenu: PropTypes.func.isRequired,
  pagination: PropTypes.object.isRequired
};

const mapStateToProps = state => ({
  meals: arraySort(state.meal.meals, 'title'),
  newMenuMeals: state.menu.newMenu,
  menus: filterify(state.menu.allMenus, state.filter),
  menuMeals: state.menu.menuMeals,
  pagination: state.menu.pagination,
  menuMealsPagination: state.menu.menuMealsPagination,
  modal: state.modal,
  editMenuMeals: state.menu.editMenu,
  mealPagination: state.meal.pagination,
});

const mapDispatchToProps = dispatch => bindActionCreators(
  {
    setNav,
    setModal,
    getMeals,
    emptyNewMenu,
    getAllMenus,
    getTodayMenu,
    updateMenu,
    postMenu,
    emptyEditMenu,
    addMealToNewMenu,
    removeMealFromNewMenu,
    setMenuForEdit,
    deleteMealInEditMenu,
    deleteMealInMenu,
    deleteMealInEditModal,
    addMealInEditMenu,
    addMealInEditMenuModal,
    getMenuMeals,
    setFilter,
    deleteMenuMeal
  },
  dispatch
);

export default connect(mapStateToProps, mapDispatchToProps)(AdminDashboard);
