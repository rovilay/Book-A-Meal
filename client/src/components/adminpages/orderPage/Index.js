/* eslint class-methods-use-this:0 */
/* eslint max-len:0 */
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';
import moment from 'moment';
import { ToastContainer, toast } from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';

import Modal from '../Modal/Index';
import navData from '../../../helpers/navData';
import tableHead from '../../../helpers/tableHead';
import TableHead from '../../common/Table/TableHead';
import OrderTableRow from '../../common/Table/OrderTableRow';
import filterAction from '../../../actions/filter';
import adminActions from '../../../actions/admin';
import FilterComp from '../../common/Filter';

class OrderHistory extends Component {
  constructor(props) {
    super(props);

    this.hideModal = this.hideModal.bind(this);
    this.showDetails = this.showDetails.bind(this);
    this.notify = this.notify.bind(this);
  }

  componentDidMount() {
    this.props.setNav(navData.adminNav);
    this.props.getAllOrders();
    this.props.filterAction('order_history', { filter: 'all' });
    this.hideModal();
  }

  hideModal() {
    this.props.setModal({
      isOpen: false,
      isEdit: false,
      isInfo: false,
      isOrderInfo: false,
      close: true,
      contentLabel: '',
    });
  }

  showDetails(orderDetails) {
    this.props.setModal({
      isOpen: true,
      isInfo: false,
      isEdit: false,
      isOrderInfo: true,
      close: false,
      contentLabel: 'Order details',
      content: { ...orderDetails }
    });
  }

  notify(msg) {
    toast(msg, {
      position: toast.POSITION.TOP_CENTER,
      className: 'toast',
      progressClassName: 'toast-progress'
    });
  }

  render() {
    const { orders, filteredOrders } = this.props;
    const sortedOrders = filteredOrders.sort((a, b) => new Date(b.createdAt) - new Date(a.createdAt));
    return (
      <div className="pull-down">
        <div className="title" id="menu-title">
          Order History
        </div>
        <hr />
        <div className="table-container">
          <FilterComp
            {...this.props}
            tableContent="order_history"
          />
          {
            (filteredOrders.length === 0)
            ?
              <p className="empty not-found">No orders found!</p>
            :
              (
                <div>
                  <table>
                    <TableHead tableHead={tableHead.orderHead} />
                    <tbody>
                      {
                        sortedOrders.map((order, i) => {
                          const {
                            id: orderId,
                            createdAt: date,
                            totalPrice,
                            deliveryAddress: address,
                            Meals: meals,
                            User
                          } = order;

                          const time = moment(date).format('HH:mm');

                          const item = {
                              sn: ++i,
                              orderId,
                              date: moment(date).format('LL'),
                              totalPrice,
                            };

                          const orderDetails = {
                            orderId,
                            meals,
                            address,
                            totalPrice,
                            time,
                            date: moment(date).format('LL'),
                            user: `${User.firstName} ${User.lastName}`
                          };

                          return (
                            <OrderTableRow
                              key={i}
                              item={item}
                              sn={++i}
                              orderDetails={orderDetails}
                              showDetails={this.showDetails}
                              notify={this.notify}
                              {...this.props}
                            />
                          );
                        })
                      }
                    </tbody>
                  </table>

                  {
                    (orders.grandTotalPrice > 0)
                    &&
                    (
                      <p>
                        Grand Total (&#8358;): {orders.grandTotalPrice}
                      </p>
                    )
                  }
                </div>
              )
          }
        </div>
        <Modal
          hideModal={this.hideModal}
          deleteRow={this.deleteRow}
          notify={this.notify}
          {...this.props}
        />
        <ToastContainer />
      </div>

    );
  }
}

OrderHistory.propTypes = {
  orders: PropTypes.object.isRequired,
  setModal: PropTypes.func.isRequired,
  modal: PropTypes.object.isRequired,
  getAllOrders: PropTypes.func.isRequired,
  filterAction: PropTypes.func.isRequired,
  filteredOrders: PropTypes.array.isRequired,
  setNav: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  orders: state.admin.orders,
  filteredOrders: state.admin.filteredOrders,
  modal: state.admin.modal,
});

const mapDispatchToProps = dispatch => bindActionCreators(
  {
    ...adminActions,
    filterAction
  },
  dispatch
);

export default connect(mapStateToProps, mapDispatchToProps)(withRouter(OrderHistory));
