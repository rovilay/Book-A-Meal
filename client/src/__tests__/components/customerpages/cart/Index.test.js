import React from 'react';
import { shallow } from 'enzyme';
import mockStore from '../../../__mockData__/mockStore';
import { Cart, mapStateToProps } from '../../../../components/customerpages/cart/Index';
import { cartDefaultState } from '../../../../reducers/cartReducer';
import CartTableRow from '../../../../components/customerpages/cart/CartTablerow';
import { orders, orderMeals } from '../../../__mockData__/orderMock';


describe('CustomerDashboard component test', () => {
  const setup = () => {
    const props = {
      cart: orders,
      cartTotalPrice: 1100,
      history: {},
      deleteMealInCart: () => jest.fn(),
      emptyCart: () => jest.fn(),
      postOrder: () => jest.fn(),
    };

    return shallow(<Cart {...props} />)
  };


  it('should render `Cart` correctly', (done) => {
    const wrapper = setup();

    expect(wrapper).toMatchSnapshot();

    done();
  });


  it('should call `onOrder`', (done) => {
    localStorage.setItem('jwt', customerToken)
    const wrapper = setup();

    const event = {
      preventDefault: () => jest.fn(),
      target: {
        name: 'email',
        value: 'test@test.com'
      }
    };

    const onOrderSpy= jest.spyOn(wrapper.instance(), 'onOrder');
    wrapper.instance().onOrder(event)
    expect(onOrderSpy).toHaveBeenCalled();

    done();
  });


  it('should call `onChange`', (done) => {
    const wrapper = setup();

    const event = {
      preventDefault: () => jest.fn(),
      target: {
        name: 'deliveryAddress',
        value: 'isuti road'
      }
    };

    const onChangeSpy= jest.spyOn(wrapper.instance(), 'onChange');
    wrapper.instance().onChange(event)
    expect(onChangeSpy).toHaveBeenCalled();

    done();
  });


  it('should call `onChange`', (done) => {
    const wrapper = setup();

    const event = {
      preventDefault: () => jest.fn(),
      target: {
        name: 'deliveryAddress',
        value: 'isuti road'
      }
    };

    const onChangeSpy= jest.spyOn(wrapper.instance(), 'onChange');
    wrapper.instance().onChange(event)
    expect(onChangeSpy).toHaveBeenCalled();

    done();
  });

  it('should call `deleteRow`', (done) => {
    const wrapper = setup();

    const deleteRowSpy= jest.spyOn(wrapper.instance(), 'deleteRow');
    wrapper.instance().deleteRow(orderMeals[0])
    expect(deleteRowSpy).toHaveBeenCalled();

    done();
  });


  it('should call `updatePortion`', (done) => {
    const wrapper = setup();

    const updatePortionSpy= jest.spyOn(wrapper.instance(), 'updatePortion');
    wrapper.instance().updatePortion(orderMeals[0])
    expect(updatePortionSpy).toHaveBeenCalled();

    done();
  });


  it('should render connected components', (done) => {
    const props = {
      item: orderMeals[0],
      removeMeal: () => jest.fn(),
      actions: {
        delete: true,
        info: false,
        edit: false
      },
      updatePortion: () => jest.fn(),
      sn: 1
    };

    const wrapper = shallow(<CartTableRow {...props}  />);
    expect(wrapper.length).toBe(1);

    done()
  });

  it('should map state to props', () => {
    const initialState = {
      cart: {
        ...cartDefaultState
      }
    };

    const ownProps = {
      cart: orderMeals,
      cartTotalPrice: 1100
    };

    const tree = mapStateToProps(initialState, ownProps);
    expect(tree).toMatchSnapshot();
  })
});
