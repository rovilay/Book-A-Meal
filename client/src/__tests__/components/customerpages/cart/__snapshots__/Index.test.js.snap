// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerDashboard component test should map state to props 1`] = `
Object {
  "cart": Array [],
  "cartTotalPrice": 0,
}
`;

exports[`CustomerDashboard component test should render \`Cart\` correctly 1`] = `
<section
  className="cartpage"
>
  <div
    className="title merienda"
    id="cart-title"
  >
    <div>
      <FontAwesome
        id="mobile-menu"
        name="cart-arrow-down"
      />
    </div>
    <h1>
      Your Cart
    </h1>
  </div>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <div
      className="address"
    >
      <label
        htmlFor="address"
      >
        Address:
      </label>
      <div
        className="input-div"
      >
        <input
          id="cart-address"
          name="deliveryAddress"
          onChange={[Function]}
          placeholder="Enter delivery address"
          required={true}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="container-test"
    >
      <div
        className="row head"
      >
        <p
          className="row-item"
          key="0"
        >
          S/N
        </p>
        <p
          className="row-item"
          key="1"
        >
          Meal
        </p>
        <p
          className="row-item"
          key="2"
        >
          Unit Price
        </p>
        <p
          className="row-item"
          key="3"
        >
          Portion
        </p>
        <p
          className="row-item"
          key="4"
        >
          Price
        </p>
        <p
          className="row-item"
          key="5"
        >
          Actions
        </p>
      </div>
      <CartTableRow
        actions={
          Object {
            "delete": true,
            "edit": false,
            "info": false,
          }
        }
        cart={
          Array [
            Object {
              "Meals": "api/v1/orders/1",
              "createdAt": "2018-08-23",
              "id": 1,
            },
            Object {
              "Meals": "api/v1/orders/2",
              "createdAt": "2018-08-24",
              "id": 2,
            },
            Object {
              "Meals": "api/v1/orders/3",
              "createdAt": "2018-08-25",
              "id": 3,
            },
          ]
        }
        cartTotalPrice={1100}
        deleteMealInCart={[Function]}
        deleteRow={[Function]}
        emptyCart={[Function]}
        history={Object {}}
        item={
          Object {
            "Meals": "api/v1/orders/1",
            "createdAt": "2018-08-23",
            "id": 1,
          }
        }
        key="0"
        postOrder={[Function]}
        showId={false}
        sn={1}
        updatePortion={[Function]}
      />
      <CartTableRow
        actions={
          Object {
            "delete": true,
            "edit": false,
            "info": false,
          }
        }
        cart={
          Array [
            Object {
              "Meals": "api/v1/orders/1",
              "createdAt": "2018-08-23",
              "id": 1,
            },
            Object {
              "Meals": "api/v1/orders/2",
              "createdAt": "2018-08-24",
              "id": 2,
            },
            Object {
              "Meals": "api/v1/orders/3",
              "createdAt": "2018-08-25",
              "id": 3,
            },
          ]
        }
        cartTotalPrice={1100}
        deleteMealInCart={[Function]}
        deleteRow={[Function]}
        emptyCart={[Function]}
        history={Object {}}
        item={
          Object {
            "Meals": "api/v1/orders/2",
            "createdAt": "2018-08-24",
            "id": 2,
          }
        }
        key="1"
        postOrder={[Function]}
        showId={false}
        sn={2}
        updatePortion={[Function]}
      />
      <CartTableRow
        actions={
          Object {
            "delete": true,
            "edit": false,
            "info": false,
          }
        }
        cart={
          Array [
            Object {
              "Meals": "api/v1/orders/1",
              "createdAt": "2018-08-23",
              "id": 1,
            },
            Object {
              "Meals": "api/v1/orders/2",
              "createdAt": "2018-08-24",
              "id": 2,
            },
            Object {
              "Meals": "api/v1/orders/3",
              "createdAt": "2018-08-25",
              "id": 3,
            },
          ]
        }
        cartTotalPrice={1100}
        deleteMealInCart={[Function]}
        deleteRow={[Function]}
        emptyCart={[Function]}
        history={Object {}}
        item={
          Object {
            "Meals": "api/v1/orders/3",
            "createdAt": "2018-08-25",
            "id": 3,
          }
        }
        key="2"
        postOrder={[Function]}
        showId={false}
        sn={3}
        updatePortion={[Function]}
      />
    </div>
    <div
      className="order"
    >
      <span
        id="order-total-price"
      >
        Total Price: â‚¦
        1100
      </span>
      <button
        className="btn-2"
        disabled={false}
        id="order-btn"
        name="orderbtn"
        type="submit"
      >
        Place order
      </button>
    </div>
  </form>
  <div
    className="empty not-found empty-cart hide"
  >
    Cart is Empty!
  </div>
</section>
`;
